Handle gzip whiteouts correctly

skaegi opened the pull request:
When looking at the resulting file contents of multilayer image we frequently are seeing `.wh.` files. We need to correctly handle whiteouts. See
1) https://github.com/opencontainers/image-spec/blob/main/layer.md#whiteouts
2) https://github.com/containerd/containerd/blob/v1.7.17/archive/tar_opts_linux.go#L34

fitzthum commented:
Might need to do `cargo fmt` on this

mkulke commented on image-rs/src/unpack.rs:
```diff
@@ -81,6 +93,63 @@ pub fn unpack<R: io::Read>(input: R, destination: &Path) -> Result<()> {
+            original_path.display()
+        ))?;
+
+        let ret = unsafe { libc::mknod(path.as_ptr(), libc::S_IFCHR, 0) };
```
can we use the nix crate for mknod?

skaegi replied:
I think we can and happy to do it. There is existing use of `libc::lutimes` and `libc::utimes` in `unpack.rs` too. @mkulke Do you want me to do this work in this PR or should I open a separate issue/PR dealing with moving to nix?

mkulke replied:
ah, wasn't aware, in this case then let's stick with the unsafe blocks here and open a dedicated issue. I think there's is good reason to use the nix crate's abstraction, but maybe there were considerations I'm not aware of

mkulke commented on image-rs/src/unpack.rs:
```diff
@@ -81,6 +93,63 @@ pub fn unpack<R: io::Read>(input: R, destination: &Path) -> Result<()> {
+            .try_into()
+            .map_err(|_| io::Error::new(io::ErrorKind::Other, "GID is too large!"))?;
+
+        let ret = unsafe { libc::lchown(path.as_ptr(), uid, gid) };
```
can we use the nix crate?

fidencio reviewed (approved):
lgtm, thanks @skaegi!

mkulke reviewed (commented):
Thank you for plugging this particular hole. It would be useful to have a test for this to ensure we don't break this during refactoring, we introduced subtle layer ordering bugs before.

skaegi commented:
> Thank you for plugging this particular hole. It would be useful to have a test for this to ensure we don't break this during refactoring, we introduced subtle layer ordering bugs before.

Sure. The existing pull image exercises this code so minimally we can add verification that the whiteouts are handled correctly.

skaegi commented:
I added whiteout tests to unpack as that seemed best and rebased on the latest. The test failure looks like an unrelated networking flake. Ready to merge…

mkulke reviewed (approved):
lgtm
